### PR TITLE
Sibling inserter: fix dead zone between blocks

### LINKED ISSUE #10648: Non-typable areas of paragraph and insertion blocks on selection
**Describe the bug**

It's easy to click areas of a paragraph or inserter block, such that they are selected, but no typing is possible

**To Reproduce**
1. Click on a large region of a paragraph or inserter block where the cursor changes from a text insert to a pointer
2. Attempt to type

**Expected behavior**

I expected what I typed to appear inside the block. It did not

**Screenshots**

https://youtu.be/mOowQS-yWXo

When a block is in this state, it is selected, but nothing can be done with it, and it's not clear that that's the case, or how to gain focus:

screenshot 2018-10-16 at 13 56 30

Additionally, the area that can be clicked on that selects but does not focus the text is rather large and easy to miss-click on

**Desktop (please complete the following information):**
 - OS: macOS
 - Browser chrome
 - Version 69

**Additional context**
- 4.0 RC1

Related to https://github.com/WordPress/gutenberg/issues/8541 which focused on just the drag handler, but this is a more general issue than that, and can be demonstrated without the drag handler

### LINKED ISSUE #18733: Block Editor: Eliminate sibling inserter dead zone
Previously: #16818

As seen in the image below, there are areas between two paragraph blocks wherein clicks do not register to focus one or the other block:

Before

In my mind, an ideal user interaction for clicks between two paragraphs is that it behave similar to continuous text of any other textarea, WYSIWYG, or word processor document; namely, that it place the text caret at the point closest to where the click occurred:

Before

Generalizing this to all blocks, I'd argue that the area between the borders of a block and its content should be treated effectively as its padding, and as with padding of any other focusable field, transition clicks into the field. The current behavior is more akin to margin applied on the block.

**Padding vs. Margin Illustration: https://codepen.io/aduth/pen/KOmLav**

This may require a rethink of how the sibling inserter behaves. Currently, there is a challenge in implementing the desired behavior because the sibling inserter is meant to be shown when hovering anywhere between two blocks, not just toward the center. If the sibling inserter is to be kept the way it is, there might need to be some implementation of a `HoverableNoPointerInteraction` component, which behaves like `pointer-events: none;` to allow a click pass-through, while still responding to `hover` events to allow the sibling inserter to be shown ([proof-of-concept](https://codepen.io/aduth/pen/JjjqQWV)).

Alternatively, we could consider a more significant rethink or removal of the sibling inserter altogether.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #18733.
Fixes #10648.

This PR removes the "dead zone" between blocks by handling clicks on the inserter to focus the closest element.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

File: packages/block-editor/src/components/block-list/insertion-point.js [hunk 1]
```diff
@@ -84,6 +89,25 @@ export default function InsertionPoint( {
 		setInserterClientId( clientId );
 	}
 
+	function onClick( event ) {
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/insertion-point.js]: For these sorts of event handlers, I'd maybe suggest to name the associated function something more meaningful to describe what we're expecting to happen as a result of the click, as a limited form of self-documentation.

e.g. `focusClosestTabbable`

File: packages/block-editor/src/components/block-list/insertion-point.js [hunk 2]
```diff
@@ -84,6 +89,25 @@ export default function InsertionPoint( {
...
+		const { clientX, clientY, target } = event;
+
+		if ( target !== ref.current ) {
```
Comment (aduth) [inline packages/block-editor/src/components/block-list/insertion-point.js]: If I'm interpreting this correctly, this exists to ensure that the click has happened _specifically_ on the wrapper, and isn't bubbled from the inserter itself (where we'd definitely not want to be transitioning the caret)?

```suggestion
		// Only handle click on the wrapper specifically, and not an event
		// bubbled from the inserter itself.
		if ( target !== ref.current ) {
```

**Review by aduth — APPROVED** [review]
I can definitely feel the improvement with this. It's a much simpler solution as well to last I explored this in #16818.

One thing I included there which I'm curious if you think would be good or not, would be to change the cursor to `cursor: text;` to indicate to the user that it's an area they can click on to start typing.

I noticed something, which I'm not sure if is expected, but if you click on the dead zone twice (above a paragraph of text), then the caret becomes lost and I can no longer type:

https://cloudup.com/cPggOJCQzqn

File: packages/block-editor/src/components/block-list/insertion-point.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/insertion-point.js]: Thanks :)

**Comment by ellatrix** [comment]
@aduth Thank you for the review! I can reproduce the issue you're having with the second click. The problem is that we don't render the sibling inserter above a block when it is selected to avoid clashing with the toolbar. I'm starting to thing we should either:

* Detect clicks on the block container instead of the inserter, and reduce the inserter click target, or;
* Always display the inserter (and maybe only hide the button if a block is selected and the inserter is displayed above the block).

The first solution sounds cleaner to me.

**Comment by ellatrix** [comment]
I'm wary to create additional event listener on the block container though, since the goal is to eventually lighten the inner blocks DOM as well. Maybe this is a good enough first version and we can revisit it separately.

**Comment by ellatrix** [comment]
Ok, I went for the second option. This also makes it more obvious if there would be a problem with the sibling inserter displaying.

**Comment by aduth** [comment]
When in navigation mode, do we expect the sibling inserter to be available?

I note that there's no way to interact with it in the latest version of the plugin.

But in `master` (presumably after these changes), it is invisible, but still exists, and can still be interacted with:

image

**Comment by aduth** [comment]
Prior to these changes, pressing Shift+Tab from the paragraph text would focus and show the sibling inserter.

It still seems to be focused, but is no longer shown.

**Comment by aduth** [comment]
Prior to these changes, there was no hover interaction "above" a block to be able to click on the sibling inserter, but now it seems to be possible (it is invisible but interactable, like in https://github.com/WordPress/gutenberg/pull/19719#issuecomment-575673499).

image

**Comment by aduth** [comment]
I see this on master as well, but noting it all the same: Pressing Shift+Tab enough times from a paragraph (going all the way through contextual toolbars and movers), eventually I see an oddly-misplaced inserter:

image

Not really sure if that's expected.

**Comment by ellatrix** [comment]
@aduth That's the expected position since #19596. I agree that this needs to be revisited, but it works fine for now. Ideally we have a tabbable inserter before and after the block that is displayed somewhere in the middle of the top and bottom border. It's also expected that the inserter is positioned before the block toolbar in the DOM.

**Comment by ellatrix** [comment]
Sorry about the visible tooltip. I'll fix it in a follow up PR.

**Pull request merged**